### DISCUSSION TITLE
Added: LDM Notes for October 27th, 2021

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#5359] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-10-27.md

## Agenda

1. UTF-8 String Literals
2. Readonly modifiers for primary constructors

**Comment by jnm2** [comment]
> We could also take primary constructors in a more radically different direction, where we'd adopt a more F#-like approach and consider these parameters captures, not fields.

This seems like a letdown. I'd still be repeating the same thing in two places instead of three. I've been looking forward to the day when this repetition would be gone. Today, much of the time that I open a .cs file, the first thing I see is similar to this:

```cs
class Foo
{
    private readonly A a;
    private readonly B b;
    private readonly C c;
    private readonly D d;

    public Foo(
        A a, // Normally these are not especially short names
        B b,
        C c,
        D d)
    {
        this.a = a;
        this.b = b;
        this.c = c;
        this.d = d;

        // Often further initialization, which is a different conversation
    }
}
```

What I was hoping for:

```cs
class Foo(
    A a,
    B b,
    C c,
    D d)
{
    // ...
}
```

> However, scoping the feature down to just primary constructor parameters was slightly less controversial. Most members of the LDM are not opposed to the concept in general, but we remain conflicted as to whether we need to have such a feature immediately. While we did get a good deal of immediate feedback, some LDM members are unsure whether this feedback will persist after users actually get their hands on the feature and give it a try.

If `readonly` can't be the default, then putting `readonly` in front of each parameter is still vastly nicer for me than usually having to say `A a`, `B b`, etc a second time as field declarations.

It would be really interesting to see if I decide that I no longer care whether implicitly declared fields are marked `readonly`. From experiencing flip-flopping between `let` and `const` for locals in TypeScript, I already know I'm not interested in the ability to declare (non-primary) parameters and locals as readonly. The scope is not that big. Maybe this can extend to fields too. It could take a while for me to find out.

What could hurt is no longer noticing if I accidentally assign to a field instead of a local in one of the methods in the class, or extra time reviewing code having to keep in mind that each field is mutable. IDE/ReSharper have nagged to make things readonly right away for so long that I've lost all feeling, so I'm not sure how often these things would bite.

**Comment by jnm2** [comment]
Is mutability a good default? Here it looks like the field is being mutated, but the constructor parameter is being mutated. And which value of `a` is assigned to `this.a`?

```cs
class Foo(int a)
{
    private int b = a++;
    
    public int B => a++;
}
```

This could be a lot more subtle with structs.
```cs
class Foo(SomeStruct s)
{
    private int x = s.SomeMutatingMethod();
    private int y = s.SomeOtherMutatingMethod();
}
```

**Comment by qrli** [comment]
> readonly is often a sensible default for locals and (in particular) parameters, but in practice LDM members do not find accidental local/parameter mutation to be a real source of bugs.

For single thread code, that's probably true. But for multi-threading code, readonly helps and give much more confidence. Well, I admit modern frameworks did a good job avoiding explicit multi-threading code in most daily work. So let's forget readonly local & parameters for now, but focus on readonly fields.

Unlike readonly local & parameter, C# already has readonly fields, for long. Refactor tools have been suggesting and auto-adding `readonly` to fields, for long. Now, if we were to have primary constructor without readonly fields, it feels like losing some feature.

**Reply by DavidArno** [reply]
Assuming it'll be possible to override the field definition:
```c#
public class C(int x)
{
    private readonly int x;
}
```
Then the likely outcome of what's being proposed by the LDM is that those refactoring tools will offer to create the override for you. 

It's disappointing that the language team seem to favour delegating this responsibility like this, but at least there would be a workaround.

**Reply by jnm2** [reply]
I expect you would need the field initializer `= x` as well, since that's how this works in records.

**Reply by DavidArno** [reply]
do you mean it likely would have to be:
```c#
public class C(int x)
{
    private readonly int x = x;
}
```
?

**Reply by CyrusNajmabadi** [reply]
We shouldn't discuss "likely" as the purpose of these designs and discussions is to decide what we actually will ship.  However, what we're discussing *now* is if it's necessary in the *initial* impl that we preview to support any sort of concept of 'readonly' or if that could, say, come in a future preview.

**Reply by jnm2** [reply]
@DavidArno Yes. If declaring the field manually, then I expect the syntax would need to be `int x = x` because this is how records work when you declare properties manually.

**Reply by qrli** [reply]
> what we're discussing now is if it's necessary in the initial impl that we preview to support any sort of concept of 'readonly' or if that could, say, come in a future preview.

My understanding is that: excluding it from initial release effectively means the "readonly by default" option has been vetoed.

**Reply by CyrusNajmabadi** [reply]
It has not been vetoed at all :-)

**Reply by DavidArno** [reply]
@CyrusNajmabadi,

There will be some people who want the fields created from a primary constructor to be read-only and there will be others that want them to be writable. In both cases, the half that "luck out" will want a way of overriding that default behaviour to get the behaviour they want.

Now you might decide in the initial release to make the default behaviour the only way of doing things, eg by not supporting `private readonly int x = x;` notation. If you do go that way, then realistically your only option is to make them writable as those that want read-only can still use the feature. Sure, it will feel rushed as it'll be missing a key piece of functionality. but we can always work around that missing functionality by using attributes and an analyzer that checks for mutation. This approach gets the feature partially done quickly, allows it to be patched by the community in the short term and dangles the carrot of doing it well at a later date in front of our noses.

Or you might decide to allow the default behaviour to be overwritten. Being able to do:
```c#
public class C(int x, int y, int z)
{
    private readonly int _y = y;
    public Z { get; set; } = z;
}
```
allows developers complete control over what happens with those primary constructor parameters from the start. There's a small chance that this will close off future development of the feature though, which has to be taken into account.

**Comment by HaloFour** [comment]
> but in practice LDM members do not find accidental local/parameter mutation to be a real source of bugs.

Until you capture them in a lambda expression in which case they become shared state where I have had to deal with bugs that result from accidental mutation.  The team was forced to adopt a breaking change in the behavior of `foreach` specifically to address this class of bugs in one case.

> We could also take primary constructors in a more radically different direction, where we'd adopt a more F#-like approach and consider these parameters captures, not fields.

I'm not 100% sure I grok this statement.  If this results in the parameter not being accessible anywhere except field initializers then you would have succeeded in partially mitigating the mutation issue (although not entirely) but at the expense of making primary constructors pretty much entirely useless.

Quite a few languages have modifiers to indicate that parameters and/or locals should be readonly, including C++.  I don't see why this is remotely controversial.  There is value in declaring intent and having it enforced by the compiler even if it doesn't otherwise affect the behavior of the code.

**Comment by alrz** [comment]
> "I fought for that when I was young and idealistic"

I can only guess what this is about. Have we given up the runtime flag to opt into having all System.String instances just become UTF-8 under the hood?

**Reply by 333fred** [reply]
Unrelated, sorry 🙂

**Comment by orthoxerox** [comment]
> We could also take primary constructors in a more radically different direction, where we'd adopt a more F#-like approach and consider these parameters captures, not fields.

The big difference is that F# record members are immutable (readonly in C# terms) by default.

**Comment by DavidArno** [comment]
> ... we took a look at another longstanding C# proposal, allowing `readonly` on parameters and locals. Several LDM members are concerned about this as a general feature, mainly because of the "attractive nuisance" issue

I'm not familiar with the term, "attractive nuisance". But a search of the term suggests it's the concept of something being attractive to children that can cause them harm. By that definition, things like mutable by default and classes being unsealed by default are examples of "attractive nuisance". But the above quote seems to suggest some of the LDM see "pit of success" features such as readonly locals as an example of "attractive nuisance", so I'm genuinely confused by this line. Have I misunderstood the term?

**Reply by jnm2** [reply]
I think it would be a mistake to read into this statement any concept related to children. The words were almost certainly used literally.

**Reply by DavidArno** [reply]
So it's a desirable feature but annoying as they don't want to do it? 😕

**Reply by qrli** [reply]
I guess that opinion comes from C/C++'s const correctness which C# had stayed away from, which adds lots of complexity but not valuable enough for routine tasks. From that point of view, it makes sense.

However, here we are talking about the simpler readonly, which has been pervasive even in Javascript for long, and it has been treated as a best practice by mainstream, with refactor and linting tools suggesting it. I think in 2021, it is not convincing to say it adds unnecessary complexity for routine tasks.

**Reply by jnm2** [reply]
@DavidArno I think the point was that it would cause an inner conflict where the same people feeling like it was desirable would be annoying themselves with it. Both an attractive force and a nuisance to the same people who are attracted. (This is not a strange concept; in nature, parasites do this all the time.)

**Reply by orthoxerox** [reply]
I think the idea is that making your parameters readonly sounds like a good idea, but gives you approximately zero actual benefit, but now 10% of your codebase is literally the word `readonly`.

**Reply by DavidArno** [reply]
> I think the idea is that making your parameters readonly sounds like a good idea, but gives you approximately zero actual benefit,

I refer you to @Richiban's analogy below.

> but now 10% of your codebase is literally the word `readonly`

That would be easily solved by allowing methods themselves to be marked `readonly` as a short-cut for marking all parameters readonly and for enforcing `let` must be used for all locals. Allowing `readonly` for parameters and `let` for readonly locals would be a necessary precursor to this though.

**Reply by orthoxerox** [reply]
An imperfect analogy is like a kitten with a all-wheel drive. I wholeheartedly agree that immutability by default is better than mutability by default for a lot of software. If there was a way to slip into a parallel universe where C# had a `mutable` keyword to mark variables, fields and parameters that you could reassign at will with the rest being read only and everything else was the same I would go there in a heartbeat. It's a small, but nice benefit.

However, I've written enough Java to get annoyed by `final`. It feels right to mark locals and parameters that you don't mutate as `final`, but my methods are short enough that it's just visual noise. I would support having `let` for locals just because it's nice and short and draws attention to that single `var` that does change, but I don't reassign my parameters unless I am replacing them with cleaned-up or default values at the very start of the method.

**Reply by HaloFour** [reply]
@orthoxerox 

> However, I've written enough Java to get annoyed by `final`.

Java has the advantage that all captures are effectively final, so the larger problem of introducing shared mutable state doesn't exist.  And yes, I have seen that catch bugs before they happen more than once.

**Reply by Przemyslaw-W** [reply]
I think that in majority of cases the intention of written code is that parameters are readonly. Sadly, this is not c# default - and that ship has already sailed. But, instead of polluting code with numerous readonly, I would actually prefer a diagnostic which I could set as compilation warning, which would flag all mutations for me. Then I could investigate such places in code and act accordingly to avoid unintended mutability bugs.

**Reply by smoothdeveloper** [reply]
F# shares same advantage as java, by default, `let` is basically same as `readonly var` we don't have in C#.

**Reply by CyrusNajmabadi** [reply]
> That would be easily solved by allowing methods themselves to be marked readonly

Methods can already be marked readonly in other contexts. It does not mean the list of parameters is readonly.  So it would be very inconsistent to use the same syntax to mean something else now.

**Reply by CyrusNajmabadi** [reply]
> I would actually prefer a diagnostic which I could set as compilation warning, which would flag all mutations for me. Then I could investigate such places in code

This is already possible today. Just write an analyzer that checks this for your own code.

**Reply by DavidArno** [reply]
> Methods can already be marked readonly in other contexts. It does not mean the list of parameters is readonly. So it would be very inconsistent to use the same syntax to mean something else now.

Good point.

**Reply by CyrusNajmabadi** [reply]
It might be possible to do something like:

`void Add(readonly: int x, int y, int z)`

though.

**Reply by qrli** [reply]
Or, enable to opt-in readonly-by-default at class level:
```cs
public readonly_by_default class Foo
{
  private Bar bar; // readonly by default
  private mutable int counter; // opt-in mutable
  public Foo(Bar bar) // parameter is readonly by default
  {
    this.bar = bar;
  }
  public void DoSomething(int a, int b, ref int c) // parameters are readonly by default, except ref/out
  { ... }
}
```

**Reply by smoothdeveloper** [reply]
I also find `readonly` modifier being the most logical, and like @CyrusNajmabadi proposal; @CyrusNajmabadi, in your example, is it only `x` which is `readonly` or the feature applies to the whole signature?

I am curious if we should expect "no change in generated IL" in code which would still compile the same after the annotation is added or if I'm already missing something there.

* should the compiler error if `readonly` is applied to `ref` or `out` parameters?
* should the compiler warn if `readonly` is applied to `in` parameter?

**Comment by Richiban** [comment]
Important note: Please don't take the following too seriously.

I find the whole readonly parameters and locals issue just maddening.

It's like the LDM is a car manufacturer and the users are asking for the next version of their cars to come with turn signals:

> Users: Can we please have turn signals on the car? That'd be great, thanks.
> 
> 	LDM: Why do you want turn signals on the car? They don't actually enable the car to do anything.
> 
> 	Users: I know, but they're really great for declaring your intent, and other road users really appreciate that. It makes it much easier for others to understand what you're doing.
> 
> 	LDM: The LDM doesn't see the value in adding turn signals. They don't actually prevent accidents.
> 
> 	Users: Well, hang on! There are definitely some senarios where they would. Plus, you can see that pretty much every other brand of car comes with turn signals so they must be a good idea.
> 
> 	LDM: We're not sure about this. The C# car is a turn signals-off-by-default vehicle, therefore we don't think we should even put turn signals on the car.
> 
> 	Users: I'm not sure I understand that argument. Again, almost every other brand of car designed in the last ten years comes with turn signals. It's starting to get frustrating that C# doesn't.
> 
> 	[Some time passes]
> 
> 	LDM: Okay, we've decided not to put turn signals on the car because we can't decide what colour they should be.
> 
> 	Users: I really don't mind what colour they are. Can we please just have them?
> 
> 	LDM: The LDM doesn't see the value in adding turn signals.
> 
> 	[Users' collective heads hit collective desks]

**Reply by HaloFour** [reply]
Because we can't have enough car analogies:

`readonly` is like seatbelts. Both are introduced much later. Both are explicitly opt-in and require additional effort on the part of the user. Both have little/no impact on operation. Both are rarely actually put to use, but in those cases when they are they can avoid catastrophic consequences. Both are worth it.

**Reply by bernd5** [reply]
Good analogy!

**Comment by TonyValenti** [comment]
I don't have insight into why a Utf8String type would be requested, but I personally think that the internal storage format of a string is an implementation detail that shouldn't be leaked.

When working with interop scenarios, I think I would prefer to see something like a [StringFormat(StringFormats.Utf8 |  StringFormats.NullTerminated)] applied to returns or method parameters and have the conversion happen automatically and have a "ToBytes" extension method that would take similar parameters to the above attribute.

**Reply by jnm2** [reply]
How can you avoid the leaking? There are some things you can represent with a single `char` that have no equivalent in UTF8, like half of a surrogate pair, and there are UTF8 characters that require more than two bytes to represent and can't fit in a `char`.

Also, string indexing/slicing is no longer random access if there is conversion going on transparently.

**Reply by CyrusNajmabadi** [reply]
> but I personally think that the internal storage format of a string is an implementation detail that shouldn't be leaked.

This would have been a nice idea to follow in .net 1. However, we're in the position caused by taking the opposite route. With strings it's very much not an internal detail. The nature of it being a contiguous sequence of 16 bit characters shows all over the place. And there is a ton of code that depends on this. 

I'm a huge proponent about allowing the runtime to change the internal representation. However, it will be massively impacting if we ever do so given the dependencies people have on the current form.

**Reply by koszeggy** [reply]
> I don't have insight into why a Utf8String type would be requested

I believe the main motivation here is the new way of JSON parsing by `System.Text.Json`, which is optimized for UTF8 strings, and has a lot of members with `ReadOnlySpan<byte>` parameters. Ironically, this new parser is blazingly fast when you parse a JSON content but only as long as you don't want to access anything by a string name, in which case it suddenly becomes [slower](https://github.com/koszeggy/KGySoft.Json#access-element-test) even than `Newtonsoft.Json`.

> I personally think that the internal storage format of a string is an implementation detail that shouldn't be leaked.

As @jnm2 also mentioned actually the UTF16 nature of `string` is also leaked. But I agree with you about that I wouldn't really like to pollute the language with a new 'alternative string' type. I would rather see an implicit conversion from a C# string literal to `ReadOnlySpan<byte>` and maybe #5295 can also help to use such strings as constants. Then it would be possible:

```cs
const ReadOnlySpan<byte> myUtf8String = "Some UTF8 string initialized from a C# string literal";
```

And then `System.Text.Json` could also work without introducing a bunch of new overloads with some new `Utf8String` type, so the already existing API could just work seamlessly.

📝 **Note:** Of course, if both `string` and `ReadOnlySpan<byte>` overloads exist, then using a string literal should pick the `string` overload so I support the idea of some `u8` postfix for this case even though I don't prefer introducing some new `Utf8String` type.

**Comment by huoyaoyuan** [comment]
> Currently, the runtime does not have a real `Utf8String` type, and instead uses `byte[]`, `Span<byte>`, or `ReadOnlySpan<byte>` as the de-facto interchange type for UTF-8 data. 

I'm interested why the runtime team doesn't accept it. The proposal and experimental implementation have existed for years.

**Reply by koszeggy** [reply]
I [mentioned](https://github.com/dotnet/csharplang/discussions/5359#discussioncomment-1560080) some reasons why I personally would not like it (confusion, unnecessary "pollution", the need of a lot of new overloads). What are your reasons to introduce such a type?

**Reply by huoyaoyuan** [reply]
A slight pain: `ReadOnlySpan<byte>` can't be used on heap, and `ReadOnlyMemory<byte>` or `byte[]` isn't as efficient/clear as `string`.
I don't use UTF8 too much though.

**Reply by koszeggy** [reply]
As for efficiency, I see it a bit differently: since we have spans, almost all public APIs have been refactored to work on `[ReadOnly]Span<char>` and even the `string` overloads are redirected to the span implementation. Such an API never needs to know whether there is a real `string` in the background. And spans are much more efficient/flexible as they can mean arrays, strings, some unmanaged memory or even stack allocated buffers. Introducing a `Utf8String` with a heap allocated array-like storage would just be a downgrade from spans, IMHO.

I risk to say that maybe we wouldn't even have `string` now if we had spans since C# 1.0. Completely new APIs often have only the span overloads that can be cast implicitly from strings anyway.

One painfully missing feature is the support of compile time constants (eg. in attributes, switch cases, parameter default values) but it would not come even for a `Utf8String` without some special handling. And hopefully #5295 will solve this issue.

> A slight pain: `ReadOnlySpan<byte>` can't be used on heap

This is true, that's why we have the `ReadOnlyMemory<byte>` and `ArraySegment<byte>` types (or the plain old array) and you can pass all of these to span parameters easily.
